capstone/capdb/models.py

```diff
@@ -1476,6 +1477,22 @@ def redact_obj(self, text):
         replacements = sorted(self.no_index_redacted.items(), reverse=True, key=lambda i: len(i[0]))
         return apply_replacements(text, replacements)
 
+    def insert_citations(self, text):
+        citations = json.loads(serializers.serialize('json', ExtractedCitation.objects.filter(cited_by_id=self.id)))
+
+        for citation in citations:
+            opinion_id = citation['fields']['opinion_id']
+            fields = citation['fields']
+            # TODO: Consider a better solution to manual casting
+            fields['target_cases'] = [int(id) for id in json.loads(fields['target_cases'])]
```

Comment:
currently, we use a Django serializer to convert the extractedcitation back into a JSON object that can enter the elasticsearch cluster. i'm curious about whether there is another method to cast that is less high touch

Reply:
Haven't tested, but something like this should work 

```
    def prepare_casebody_data(self, instance):
        body = instance.body_cache
        text = deepcopy(body.json)
        serializer = self._fields['casebody_data']['text']['opinions']['extracted_citations']
        cites_by_id = group_by(instance.extracted_citations.all(), lambda c: c.opinion_id)
        for i, opinion in text['opinions']:
            if i in cites_by_id:
                opinion['extracted_citations'] = [serializer.get_value_from_instance(c) for c in cites_by_id[i]]
        return instance.redact_obj({
            'xml': body.xml,
            'html': body.html,
            'text': text,
        })
```

Note by the way that `ExtractedCitation.objects.filter(cited_by_id=self.id)` wouldn't be performant here -- indexing a volume in elasticsearch should take `O(1)` sql queries, which is handled by a `prefetch_related('extracted_citations')` hiding in `update_elasticsearch_for_vol`. That means we get `instance.extracted_citations.all()` for free, but any other queries to get the same data would hit the DB.

Reply:
thanks! i'm snagging something like this to also filter on the `cited_by_id` without eating a second query against the db:
```
    def prepare_casebody_data(self, instance):
        body = instance.body_cache
        serializer = self._fields['casebody_data']['text']['opinions']['extracted_citations']
        # We must unroll to dict to support the `in` operator
        outbound_cites = [c for c in ExtractedCitation.objects.all() if c.cited_by_id == instance.id]
        cites_by_id = {k: list(v) for k,v in groupby(outbound_cites, lambda c: c.opinion_id)}

        for i, opinion in enumerate(body.json['opinions']):
            if i in cites_by_id:
                body.json['opinions'][i]['extracted_citations'] = [serializer.get_value_from_instance(c) for c in cites_by_id[i]]

        return instance.redact_obj({
            'xml': body.xml,
            'html': body.html,
            'text': body.json,
        })
```